### PR TITLE
fix: resolve mentionable page size from config, not stale room counts

### LINKED ISSUE #419: [Fix] subscription.mentionable returns empty on stale counts — drop the cap check, use env limits
> **Verified against:** `main`@c604d0cf — code read after a fresh `git pull` (code is source of truth over docs).

## Summary
`subscription.mentionable` returns an **empty list for a genuinely populated room** when the room's denormalized `userCount`/`appCount` are stale or unset. The handler derives its limit — and an early empty-return — from `cap = room.userCount + room.appCount`, so wrong counts fake an empty room even though the subscriptions exist. Fix: stop deriving anything from the counts; use env-based limits and query the subscriptions directly.

## Reproduce steps
- **Preconditions:** a channel with ≥2 members (users and/or bots) whose room doc has `userCount = 0` and `appCount = 0` (a count that never got set, drifted, or a cross-site replica).
- **Steps:**
  1. A member calls `chat.user.{account}.request.room.{roomID}.{siteID}.subscription.mentionable`.
- **Expected vs actual:** Expected — the other members are returned. Actual — an **empty list**, because `cap == 0` short-circuits before the query.
- **Evidence (dev stack):** confirmed the empty-list path is the `cap == 0` early-return (`handler.go:625-627`) reading the denormalized counts; a populated room with those counts at 0 returns empty despite real subscriptions.

## Root cause  (verified, file:line)
- `listMentionableSubscriptions` (`room-service/handler.go:602`): `mentionableCap = room.UserCount + room.AppCount`; if `cap == 0` it returns an empty list immediately (`:625-627`) — **before** querying `subscriptions`. It also uses `cap` for the limit: no-limit → `min(3, cap)` (`:631`), explicit → `min(limit, cap)` (`:637`).
- Those counts are the **denormalized room-doc fields** (read via `GetRoom`), not a live count — same denorm-drift family as the restricted-inherit fix (#418) and the cross-site count ownership. When they're wrong, mentionable is wrong.

## Message-path trace
1. Client request struct + docs/client-api.md — no change (`{limit?, filter?}`).
2. Gatekeeper — N/A (read path).
3. Canonical model / event structs — no change.
4. Workers — N/A.
5. Storage — no change; the store query already `$limit`s the real subscription rows.
6. Read path (room-service) — **change here**: drop the `cap` dependency; env limits; query directly.
7. Client-facing events — N/A.

## Fix  (Primary — agreed with mliu33)
- **Remove the `cap == 0` early-return.** An empty room's store query already returns `[]`; the early-return only fakes emptiness on bad counts.
- **Replace the count-derived limit with two env knobs** (in room-service config):
  - `MENTIONABLE_DEFAULT_LIMIT` (default 3) — used when the client sends no `limit` (preserves today's preview size).
  - `MENTIONABLE_MAX_LIMIT` (default e.g. 50) — an explicit `limit` clamps to `min(limit, MAX)` instead of to `cap`.
- Keep `explicit limit <= 0 → reject` (`errMentionableLimitInvalid`).
- **Guardrail — never let the effective `$limit` be 0** (Mongo `$limit: 0` errors "limit must be positive"): validate both env values `> 0` at startup, and ensure the resolved limit is always ≥ 1. This is the invariant the old `cap == 0` guard was accidentally covering.
- **Simplification:** after removing `cap`, the handler no longer needs the room doc — only the membership check. If `room` is unused elsewhere in the handler, replace `requireMembershipAndGetRoom` with a plain `CheckMembership` and drop the `GetRoom` round-trip (verify no other use first).
- **Alternative considered — pagination:** rejected for now. Response bytes are already bounded by `boundedReply`, and a mention picker is type-to-filter (the FE narrows via `filter`); pagination is more machinery than the surface needs. Add later only if a concrete large-room need appears.

## Contracts & docs
- docs/client-api.md updated same PR? Y if the limit semantics are documented there (the over-cap behavior note); otherwise N.
- Schema change? N. New **config** only: two optional env vars with defaults (deploy note, not a DB migration).

## Tests
- TDD red→green: a `listMentionableSubscriptions` test where the room's `userCount`/`appCount` are 0 but subscriptions exist → the members are returned (fails today via the early-return, passes after).
- Env-limit tests: no-limit → `MENTIONABLE_DEFAULT_LIMIT` rows; explicit over-max → clamped to `MAX`; `limit <= 0` → reject; the effective `$limit` is never 0.
- **e2e (dev stack):** required — force a populated room's counts to 0 and assert mentionable now returns the members (harness pattern already exists: `mentionable-user-role` / `repro-mentionable-limit`).

## Acceptance criteria
- A populated room returns its mentionable members regardless of the denormalized `userCount`/`appCount`.
- No-limit uses the env default; explicit limit clamps to the env max; `limit <= 0` rejects; the query never issues `$limit: 0`.

## Out of scope / open questions
- The **root cause of the counts not being set** (why `userCount`/`appCount` are 0/stale) — a separate concern; those counts also feed the add-member capacity check and the sidebar. Tracked separately.
- Final default values for the two env vars (3 / 50 are placeholders — mliu33's call).

## PR / merge
- pr-self-audit is an internal pre-review gate; not in the PR body.
- Source repo `hmchangw/newchat`: external review + external merge — open/rebase/CI-green/request-review; do not solo-merge.

### PR DESCRIPTION
**Pull request by chenjr0719**
## Summary

`subscription.mentionable` returned an empty list for a populated room when the room's denormalized `userCount`/`appCount` were stale or `0`. The handler derived both an early empty-return and its page-size limit from `cap = room.UserCount + room.AppCount`, so a `0` cap short-circuited to `[]` and `@`-mention autocomplete showed nothing — even though the room had members.

Fixes #419.

## Changes

- **room-service/handler.go** — `listMentionableSubscriptions` no longer reads the room's member counts:
  - Removed the `cap == 0` early-return and the count-based limit clamp.
  - Resolve the page size from config instead: nil request-limit → `MENTIONABLE_DEFAULT_LIMIT`; explicit limit `<= 0` → rejected; otherwise clamped to `MENTIONABLE_MAX_LIMIT`.
  - Gate on membership alone via `CheckMembership` and drop the `GetRoom` round-trip (the room document was only used for the counts). The parallel `requireMembershipAndGetRoom` helper is retained — `listMemberStatuses` still uses it.
  - Removed the now-unused `defaultMentionableLimit` const.
- **room-service/main.go** — added `MENTIONABLE_DEFAULT_LIMIT` (default `3`) and `MENTIONABLE_MAX_LIMIT` (default `50`) to `config`, each validated `> 0` at startup (mirrors the existing `RESTRICTED_ROOM_MIN_MEMBERS` fail-fast), wired onto the handler.
- **room-service/helper.go** — `errMentionableLimitInvalid` message narrowed to `"limit must be > 0"` (it no longer references the room counts).
- **docs/client-api.md** + **docs/client-api/request-reply.md** — updated the `limit` semantics and the error string.
- Tests updated/added (see below).

The store method `ListMentionableSubscriptions` already applies `$limit` to the real subscription rows, so a populated room now returns its members regardless of the denormalized counts. The resolved limit is always `>= 1` (startup validation guarantees the bounds are `> 0`).

## Verification

- Red→green TDD: the mentionable handler tests were rewritten for the count-independent design and fail to compile against the old code (they reference the new config/handler fields); they pass after the change.
- `go test ./room-service/...` — pass (unit).
- `go vet ./room-service/` and `go vet -tags integration ./room-service/` — clean.
- `golangci-lint run ./room-service/...` — 0 issues.
- Key tests:
  - `TestHandler_ListMentionableSubscriptions/stale_zero_counts_still_returns_members` — the #419 regression: a populated room with no `GetRoom` call still returns members.
  - `TestHandler_ListMentionableSubscriptions/explicit_limit_over_max_clamps_to_max`, `.../limit_zero_rejected_before_store_call`, `.../limit_negative_rejected_before_store_call`, `.../default_limit,_empty_filter,_happy_path` — env-limit resolution + reject/clamp, store never called with limit 0.
  - `TestConfig_MentionableLimits` — the two env vars parse (defaults + override) and the `> 0` startup guard rejects non-positive values.

## Summary by CodeRabbit

- **New Features**
  - Added configurable default and maximum page sizes for mention autocomplete.
  - Pagination limits now operate independently of room membership counts.

- **Bug Fixes**
  - Positive limits are accepted and capped at the configured maximum.
  - Invalid pagination settings are rejected during service startup.

- **Documentation**
  - Updated API documentation and error descriptions to reflect the revised pagination behavior.

### CODE REVIEW

**Comment by julianshen** [comment]
**4 findings · 4 inline comments posted** — 🟠 4 warning

### Related work

- [Issue #419](https://github.com/hmchangw/newchat/issues/419) — \[Fix\] subscription.mentionable returns empty on stale counts — drop the cap check, use env limits (open)

<details>
<summary>📒 Files reviewed (7)</summary>

* `docs/client-api.md`
* `docs/client-api/request-reply.md`
* `room-service/config_test.go`
* `room-service/handler.go`
* `room-service/handler_test.go`
* `room-service/helper.go`
* `room-service/main.go`

</details>
<details>
<summary>⚙️ Rules run (11)</summary>

* `cassandra`
* `dead-code`
* `distributed-system`
* `elasticsearch`
* `mongodb`
* `nats-large-payload`
* `nats`
* `oom`
* `performance-and-concurrency`
* `silent-failure`
* `tgd-review`

</details>

---

_🤖 Posted by [tGDBot](https://github.com/julianshen/tGDBot) using `openai-codex/gpt-5.6-sol`_

**Review by julianshen — COMMENTED** [review]
### tGDBot review of `43f00d6` — 4 findings in 7 files

4 🟠 warning
Findings not shown inline are in the [review summary](https://github.com/hmchangw/newchat/pull/420#issuecomment-5456981725), which is authoritative for what this run found and where each finding went.
Rules: `cassandra`, `dead-code`, `distributed-system`, `elasticsearch`, `mongodb`, `nats-large-payload`, `nats`, `oom` +3 more

<details>
<summary>How to read an inline comment</summary>

- Each one opens with `category | severity | effort | rule`.
- Severity is one of 🔴 Blocking · 🟠 Warning · 🔵 Suggestion.
- Effort is ⚡ Quick fix or 🏗️ Heavy lift — the SIZE of the change, not its importance.
- A suggestion block is committable, and applies verbatim — read it first, it is generated text.
- The collapsed 🤖 block is a ready-made prompt for handing that one finding to a coding agent.

</details>

<details>
<summary>Replying to @julianshen</summary>

In a finding's own thread: `@julianshen explain`, `@julianshen reconsider <reason>`.

Anywhere on the pull request: `@julianshen check latest`, `@julianshen review focus: <direction>`,
`@julianshen remember <lesson>`, `@julianshen memories`, `@julianshen forget <memory-id>`.

One command per comment.

</details>

---

_🤖 Posted by [tGDBot](https://github.com/julianshen/tGDBot) using `openai-codex/gpt-5.6-sol`_

File: room-service/main.go [hunk 1]
```diff
@@ -170,6 +175,14 @@ func main() {
 		slog.Error("invalid RESTRICTED_ROOM_MIN_MEMBERS: must be > 0", "value", cfg.RestrictedRoomMinMembers)
 		os.Exit(1)
 	}
+	if cfg.MentionableDefaultLimit <= 0 {
+		slog.Error("invalid MENTIONABLE_DEFAULT_LIMIT: must be > 0", "value", cfg.MentionableDefaultLimit)
+		os.Exit(1)
+	}
+	if cfg.MentionableMaxLimit <= 0 {
```
Comment (julianshen) [inline room-service/main.go]: _🏷️ backpressure_ | _🟠 Warning_ | _⚡ Quick fix_ | _`distributed-system`_

**The configured maximum does not cap the default page size**

Startup accepts MENTIONABLE_DEFAULT_LIMIT greater than MENTIONABLE_MAX_LIMIT. For example, with default=100000 and max=50, a request without `limit` sends 100000 to ListMentionableSubscriptions, bypassing the configured safety cap. Repeated autocomplete requests against a large room can therefore amplify database work and response size despite the operator setting a maximum. Reject this inconsistent configuration at startup (and add a corresponding config test), or clamp the default to the maximum.

<details>
<summary>📝 Committable suggestion</summary>

> ‼️ **IMPORTANT**
> Review this carefully before committing. It is generated from an automated
> review of an untrusted diff — make sure it replaces exactly the highlighted
> lines, is complete, is correctly indented, and actually does what you want.

```suggestion
	if cfg.MentionableMaxLimit <= 0 {
		slog.Error("invalid MENTIONABLE_MAX_LIMIT: must be > 0", "value", cfg.MentionableMaxLimit)
		os.Exit(1)
	}
	if cfg.MentionableDefaultLimit > cfg.MentionableMaxLimit {
		slog.Error("invalid MENTIONABLE_DEFAULT_LIMIT: must be <= MENTIONABLE_MAX_LIMIT", "default", cfg.MentionableDefaultLimit, "max", cfg.MentionableMaxLimit)
		os.Exit(1)
	}
```

</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `room-service/main.go` around line 179: Startup accepts MENTIONABLE_DEFAULT_LIMIT greater than MENTIONABLE_MAX_LIMIT. For example, with default=100000 and max=50, a request without `limit` sends 100000 to ListMentionableSubscriptions, bypassing the configured safety cap. Repeated autocomplete requests against a large room can therefore amplify database work and response size despite the operator setting a maximum. Reject this inconsistent configuration at startup (and add a corresponding config test), or clamp the default to the maximum.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

---

_🤖 Posted by [tGDBot](https://github.com/julianshen/tGDBot) using `openai-codex/gpt-5.6-sol`_

File: room-service/handler.go [hunk 2]
```diff
@@ -617,24 +620,25 @@ func (h *Handler) listMentionableSubscriptions(c *natsrouter.Context) (*model.Me
...
-	var limit int
-	if req.Limit == nil {
-		limit = min(defaultMentionableLimit, mentionableCap)
-	} else {
-		limit = *req.Limit
-		if limit <= 0 {
+	// Startup validation guarantees both bounds are > 0, so the resolved limit is
+	// always >= 1: a nil request-limit yields the default, an explicit one is
+	// rejected when <= 0 and otherwise clamped to the configured max.
+	limit := h.mentionableDefaultLimit
```
Comment (julianshen) [inline room-service/handler.go:636]: _🏷️ nats-large-payload_ | _🟠 Warning_ | _⚡ Quick fix_ | _`nats-large-payload`_

**Config limits allow oversized mentionable-subscription replies**

The NATS request/reply handler resolves its query size directly from `mentionableDefaultLimit` or `mentionableMaxLimit`, while startup only validates that these values are positive. A valid configuration such as `MENTIONABLE_DEFAULT_LIMIT=10000` followed by a request without `limit` returns up to 10,000 `model.MentionableSubscription` rows in one `model.MentionableSubscriptionsResponse`. The documented row fields include `account`, `engName`, `chineseName`, `app.name`, and `app.assistant.name`; even user rows containing only the serialized `account` field exceed 128 KB at that count. The router then serializes the entire response into the NATS sync reply, causing unnecessary allocation, serialization latency, and consumer memory pressure. Enforce a fixed safe upper bound on both configured values (and require default <= max), or paginate/chunk the response rather than allowing an unbounded configured page size.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `room-service/handler.go` around line 636: The NATS request/reply handler resolves its query size directly from `mentionableDefaultLimit` or `mentionableMaxLimit`, while startup only validates that these values are positive. A valid configuration such as `MENTIONABLE_DEFAULT_LIMIT=10000` followed by a request without `limit` returns up to 10,000 `model.MentionableSubscription` rows in one `model.MentionableSubscriptionsResponse`. The documented row fields include `account`, `engName`, `chineseName`, `app.name`, and `app.assistant.name`; even user rows containing only the serialized `account` field exceed 128 KB at that count. The router then serializes the entire response into the NATS sync reply, causing unnecessary allocation, serialization latency, and consumer memory pressure. Enforce a fixed safe upper bound on both configured values (and require default <= max), or paginate/chunk the response rather than allowing an unbounded configured page size.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

---

_🤖 Posted by [tGDBot](https://github.com/julianshen/tGDBot) using `openai-codex/gpt-5.6-sol`_

File: room-service/main.go [hunk 3]
```diff
@@ -170,6 +175,14 @@ func main() {
...
+		slog.Error("invalid MENTIONABLE_MAX_LIMIT: must be > 0", "value", cfg.MentionableMaxLimit)
+		os.Exit(1)
+	}
 	roomRouteMode, err := subject.ParseRoomRouteMode(cfg.RoomSubjectMode)
```
Comment (julianshen) [inline room-service/main.go]: _🎯 correctness_ | _🟠 Warning_ | _⚡ Quick fix_ | _`tgd-review`_

**Startup accepts a default limit larger than the maximum**

Only positivity is validated, so `MENTIONABLE_DEFAULT_LIMIT=100` with `MENTIONABLE_MAX_LIMIT=50` starts successfully. A request without a limit then queries with 100, while explicit limits are capped at 50, defeating the configured page-size bound. Reject configurations where the default exceeds the maximum and cover that case in config tests.

<details>
<summary>📝 Committable suggestion</summary>

> ‼️ **IMPORTANT**
> Review this carefully before committing. It is generated from an automated
> review of an untrusted diff — make sure it replaces exactly the highlighted
> lines, is complete, is correctly indented, and actually does what you want.

```suggestion
	if cfg.MentionableDefaultLimit <= 0 {
		slog.Error("invalid MENTIONABLE_DEFAULT_LIMIT: must be > 0", "value", cfg.MentionableDefaultLimit)
		os.Exit(1)
	}
	if cfg.MentionableMaxLimit <= 0 {
		slog.Error("invalid MENTIONABLE_MAX_LIMIT: must be > 0", "value", cfg.MentionableMaxLimit)
		os.Exit(1)
	}
	if cfg.MentionableDefaultLimit > cfg.MentionableMaxLimit {
		slog.Error("invalid mentionable limits: MENTIONABLE_DEFAULT_LIMIT must be <= MENTIONABLE_MAX_LIMIT", "default", cfg.MentionableDefaultLimit, "max", cfg.MentionableMaxLimit)
		os.Exit(1)
	}
```

</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `room-service/main.go` around line 179: Only positivity is validated, so `MENTIONABLE_DEFAULT_LIMIT=100` with `MENTIONABLE_MAX_LIMIT=50` starts successfully. A request without a limit then queries with 100, while explicit limits are capped at 50, defeating the configured page-size bound. Reject configurations where the default exceeds the maximum and cover that case in config tests.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

---

_🤖 Posted by [tGDBot](https://github.com/julianshen/tGDBot) using `openai-codex/gpt-5.6-sol`_

File: room-service/config_test.go [hunk 4]
```diff
@@ -94,6 +94,48 @@ func TestConfig_ValkeyAddrsParsed(t *testing.T) {
...
+		t.Setenv("MENTIONABLE_MAX_LIMIT", "200")
+		cfg, err := env.ParseAs[config]()
+		require.NoError(t, err)
+		assert.Equal(t, 10, cfg.MentionableDefaultLimit)
+		assert.Equal(t, 200, cfg.MentionableMaxLimit)
+	})
+
+	// The > 0 startup gate (fail-fast in main) rejects a non-positive value; env
+	// parsing itself accepts it, so the guard is what must catch it.
+	t.Run("non-positive rejected by startup guard", func(t *testing.T) {
```
Comment (julianshen) [inline room-service/config_test.go]: _🧪 tests_ | _🟠 Warning_ | _⚡ Quick fix_ | _`tgd-review`_

**Non-positive config test never exercises startup validation**

This subtest only parses the values and verifies the tautologies `!(0 > 0)` and `!(-1 > 0)`; it would still pass if either startup guard in `main.go` were deleted or inverted. Extract the startup validation into a testable helper and assert that each non-positive configuration is rejected.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `room-service/config_test.go` around line 129: This subtest only parses the values and verifies the tautologies `!(0 > 0)` and `!(-1 > 0)`; it would still pass if either startup guard in `main.go` were deleted or inverted. Extract the startup validation into a testable helper and assert that each non-positive configuration is rejected.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

---

_🤖 Posted by [tGDBot](https://github.com/julianshen/tGDBot) using `openai-codex/gpt-5.6-sol`_

**Comment by chenjr0719** [comment]
Thanks — these are valid, holding this before merge to address them:

1. **`default` can exceed `max`** (main.go): startup validates each value `> 0` but not `default <= max`, so a no-`limit` request can query with a default that bypasses the configured maximum. Fix: reject `default > max` at startup (extracted into a testable helper).
2. **Unbounded reply** (handler.go): the mentionable reply is returned raw, not through `boundedReply` (unlike `app.cmd-menu`/`app.tabs`), so a large configured limit yields an oversized NATS reply. Fix: wrap the response in `boundedReply` for a byte cap, and bound the limits per (1).
3. **Non-positive config test is tautological** (config_test.go): it asserts `!(0>0)` and doesn't exercise the startup guard. Fix: assert rejection through the extracted validation helper.

Reworking now — will re-request review. Holding merge until then.

**Review by mliu33 — APPROVED** [review]
Thanks a lot!
